ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/bounded/BoundedQueue.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.manager.bounded;
+
+import tech.pegasys.pantheon.metrics.Counter;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.metrics.PantheonMetricCategory;
+
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+public class BoundedQueue extends LinkedBlockingDeque<Runnable> {
+  public static final long DEFAULT_EVICTION_POLL_TIMEOUT = 10;
+  public static final TimeUnit DEFAULT_EVICTION_POLL_TIMEOUT_UNIT = TimeUnit.SECONDS;
+
+  private final long evictionPollTimeout;
+  private final TimeUnit evictionPollTimeoutUnit;
+  private final MetricsSystem metricsSystem;
+  private final Counter totalEvictedTaskCounter;
+
+  public BoundedQueue(
+      final int capacity,
+      final long evictionPollTimeout,
+      final TimeUnit evictionPollTimeoutUnit,
+      final String metricName,
+      final MetricsSystem metricsSystem) {
+    super(capacity);
+    this.evictionPollTimeout = evictionPollTimeout;
+    this.evictionPollTimeoutUnit = evictionPollTimeoutUnit;
+    this.metricsSystem = metricsSystem;
+    this.totalEvictedTaskCounter =
+        this.metricsSystem.createCounter(
+            PantheonMetricCategory.EXECUTORS,
+            metricName + "_dropped_tasks_total",
+            "Total number of tasks rejected by this working queue.");
+  }
+
+  @Override
+  public boolean offer(final Runnable task) {
+    try {
+      if (!super.offer(task)) {
+        poll(evictionPollTimeout, evictionPollTimeoutUnit);
```

Comment:
We shouldn't need to use `poll` and have a timeout here.  If the queue is full there must be an item to remove and if there isn't something is very, very wrong.  So we can just use `remove()` and get rid of the related config options.

There's also a race condition here - if two threads are calling `offer` at the same time, one may be rejected, remove an item to make room but then the other thread fills the new spot, which results in this one being rejected incorrectly.

I suspect this is the place where we want to use a tight loop - as long as `capacity` is greater than 0, eventually we'll be able to remove an item and fill the available slot before another thread does.  So something like:

```
while (!super.offer(task)) {
  remove();
}
return true;
```

Reply:
Ok it makes sense. Done.